test/devkit_test.go

```diff
@@ -0,0 +1,164 @@
+	"openpitrix.io/openpitrix/pkg/utils/idtool"
+	"openpitrix.io/openpitrix/test/client/app_manager"
+	"openpitrix.io/openpitrix/test/client/repo_indexer"
+	"openpitrix.io/openpitrix/test/client/repo_manager"
+	"openpitrix.io/openpitrix/test/models"
+)
+
+var testRepoDir = path.Join("/tmp/openpitrix-test", idtool.GetUuid(""))
+
+func execOnTestRepo(t *testing.T, cmd string) string {
+	fullCmd := fmt.Sprintf("docker run --rm -i --name='test-op' -p 8879:8879 -v %s:/tmp/openpitrix-test -w /tmp/openpitrix-test openpitrix %s", testRepoDir, cmd)
+	t.Logf("run command [%s]", fullCmd)
+	c := exec.Command("/bin/sh", "-c", fullCmd)
+	output, err := c.CombinedOutput()
+	assert.NoError(t, err)
+	return string(output)
+}
+
+func execOnTestRepoD(t *testing.T, cmd string) string {
+	fullCmd := fmt.Sprintf("docker run --rm -i -d --name='test-op' -p 8879:8879 -v %s:/tmp/openpitrix-test -w /tmp/openpitrix-test openpitrix %s", testRepoDir, cmd)
+	t.Logf("run command [%s]", fullCmd)
+	c := exec.Command("/bin/sh", "-c", fullCmd)
+	output, err := c.CombinedOutput()
+	assert.NoError(t, err)
+	return string(output)
+}
+
+func cleanupTestRepoDocker(t *testing.T) string {
+	output, err := exec.Command("/bin/sh", "-c", "docker rm -f test-op").CombinedOutput()
+	assert.NoError(t, err)
+	return string(output)
+}
+
+func TestDevkit(t *testing.T) {
+	t.Logf("start create repo at [%s]", testRepoDir)
+
+	t.Log(execOnTestRepo(t, "op create nginx"))
+
+	t.Log(execOnTestRepo(t, "ls nginx"))
+
+	t.Log(execOnTestRepo(t, "op package nginx"))
+
+	t.Log(execOnTestRepo(t, "op index ./"))
+
+	t.Log(execOnTestRepo(t, "cat index.yaml"))
+
+	ip := strings.TrimSpace(execOnTestRepo(t, "hostname -i"))
+	localIp := GetLocalIP()
+	t.Log(execOnTestRepoD(t, fmt.Sprintf("op serve --address %s:8879 --url http://%s:8879/", ip, localIp)))
+
+	t.Run("create repo", func(t *testing.T) {
+		time.Sleep(5 * time.Second)
+		testCreateRepo(t)
+	})
+
+	// cleanup
+	t.Log(cleanupTestRepoDocker(t))
+	t.Log(execOnTestRepo(t, "find . -mindepth 1 -delete"))
+}
+
+func waitRepoEventSuccess(t *testing.T, repoEventId string) {
+	client := GetClient(clientConfig)
+
+	for {
+		describeEventParams := repo_indexer.NewDescribeRepoEventsParams()
+		describeEventParams.RepoEventID = []string{repoEventId}
+		describeEventResp, err := client.RepoIndexer.DescribeRepoEvents(describeEventParams)
+		assert.NoError(t, err)
+		assert.Equal(t, int64(1), describeEventResp.Payload.TotalCount, "count should be 1")
+		assert.Equal(t, repoEventId, describeEventResp.Payload.RepoEventSet[0].RepoEventID, "error repo event id")
+
+		status := describeEventResp.Payload.RepoEventSet[0].Status
+		assert.NotEqual(t, constants.StatusFailed, status, "status should not be failed")
+
+		switch status {
+		case constants.StatusSuccessful:
+			return
+		case constants.StatusPending, constants.StatusWorking:
+			time.Sleep(5 * time.Second)
+		}
+		continue
+	}
+}
+
+// GetLocalIP returns the non loopback local IP of the host
+func GetLocalIP() string {
```

Comment:
There exist one such func in service/metadata/drone/utils,  can the two be merged into one?

Reply:
done